### PR TITLE
Evaluate discrete, first cut without hinting

### PR DESCRIPTION
**Pull request by eggrobin**
`DiscreteTrajectory` is now a `Trajectory`. The existing use of `Hermite3` is also for interpolation of discrete trajectories (when computing apsides in `Ephemeris`), we may want to switch to `Evaluate` there too in a future CL.

### CODE REVIEW

File: physics/discrete_trajectory_body.hpp [hunk 1]
```diff
@@ -178,6 +183,110 @@ void DiscreteTrajectory<Frame>::ForgetBefore(Instant const& time) {
...
+    CHECK_LE(t_min(), time);
+    CHECK_GE(t_max(), time);
+    // This is the upper bound of the interval upon which we will do the
+    // interpolation.
+    auto const upper = LowerBound(time);
+    if (upper == Begin()) {
+      return upper.degrees_of_freedom().position();
+    }
+    auto const lower = --Iterator{upper};
+    Hermite3<Instant, Position<Frame>> interpolation{
```
Comment (pleroy) [inline physics/discrete_trajectory_body.hpp]: const

File: physics/discrete_trajectory_body.hpp [hunk 2]
```diff
@@ -178,6 +183,110 @@ void DiscreteTrajectory<Frame>::ForgetBefore(Instant const& time) {
...
+        {lower.time(), upper.time()},
+        {lower.degrees_of_freedom().position(),
+         upper.degrees_of_freedom().position()},
+        {lower.degrees_of_freedom().velocity(),
+         upper.degrees_of_freedom().velocity()}};
+    return {interpolation.Evaluate(time),
+            interpolation.EvaluateDerivative(time)};
```
Comment (pleroy) [inline physics/discrete_trajectory_body.hpp]: Could we please factor out this triplication by having a function give-me-the-Hermite-polynomial for the time and the hint?

File: physics/discrete_trajectory_test.cpp [hunk 3]
```diff
@@ -815,6 +824,45 @@ TEST_F(DiscreteTrajectoryTest, IteratorSuccess) {
...
+    EXPECT_THAT(circle.EvaluatePosition(t0_ + t, /*hint=*/nullptr),
+                Eq(q_interpolated));
+    EXPECT_THAT(circle.EvaluateVelocity(t0_ + t, /*hint=*/nullptr),
+                Eq(v_interpolated));
+    max_r_error = std::max(
+        max_r_error, RelativeError(r, (q_interpolated - World::origin).Norm()));
+    max_v_error =
+        std::max(max_v_error, RelativeError(v, v_interpolated.Norm()));
+  }
+  // The actual error is about 1.5 %.
```
Comment (pleroy) [inline physics/discrete_trajectory_test.cpp]: Then make it 1.4%..1.6% or some such.

File: physics/discrete_trajectory_test.cpp [hunk 4]
```diff
@@ -815,6 +824,45 @@ TEST_F(DiscreteTrajectoryTest, IteratorSuccess) {
...
+  EXPECT_THAT(max_r_error, AllOf(Ge(0.01), Le(0.02)));
+  // The actual error is about 1.2 %.
+  EXPECT_THAT(max_v_error, AllOf(Ge(0.01), Le(0.02)));
```
Comment (pleroy) [inline physics/discrete_trajectory_test.cpp]: Same.

File: physics/discrete_trajectory_body.hpp (hunk 2 above)
Comment (eggrobin) [inline physics/discrete_trajectory_body.hpp]: Done

File: physics/discrete_trajectory_body.hpp (hunk 1 above)
Comment (eggrobin) [inline physics/discrete_trajectory_body.hpp]: OBE.

File: physics/discrete_trajectory_test.cpp (hunk 3 above)
Comment (eggrobin) [inline physics/discrete_trajectory_test.cpp]: Done

File: physics/discrete_trajectory_test.cpp (hunk 4 above)
Comment (eggrobin) [inline physics/discrete_trajectory_test.cpp]: Done

File: physics/discrete_trajectory_body.hpp [hunk 5]
```diff
@@ -209,20 +206,7 @@ Position<Frame> DiscreteTrajectory<Frame>::EvaluatePosition(
     // TODO(egg): do things with the hint.
     CHECK_LE(t_min(), time);
     CHECK_GE(t_max(), time);
```
Comment (pleroy) [inline physics/discrete_trajectory_body.hpp]: The checks should be in `GetInterpolation`.

File: physics/discrete_trajectory_body.hpp [hunk 6]
```diff
@@ -209,20 +206,7 @@ Position<Frame> DiscreteTrajectory<Frame>::EvaluatePosition(
     // TODO(egg): do things with the hint.
```
Comment (pleroy) [inline physics/discrete_trajectory_body.hpp]: Remove.

File: physics/discrete_trajectory_body.hpp [hunk 7]
```diff
@@ -209,20 +206,7 @@ Position<Frame> DiscreteTrajectory<Frame>::EvaluatePosition(
...
-    }
-    auto const lower = --Iterator{upper};
-    Hermite3<Instant, Position<Frame>> interpolation{
-        {lower.time(), upper.time()},
-        {lower.degrees_of_freedom().position(),
-         upper.degrees_of_freedom().position()},
-        {lower.degrees_of_freedom().velocity(),
-         upper.degrees_of_freedom().velocity()}};
-    return interpolation.Evaluate(time);
+    return GetInterpolation(time, hint).Evaluate(time);
```
Comment (pleroy) [inline physics/discrete_trajectory_body.hpp]: You could use `down_cast_hint` here and avoid the block altogether.

**Pull request merged**